cache/async_queue_test.go

```diff
@@ -0,0 +1,51 @@
+package cache
+
+import (
+	"testing"
+	"time"
+
+	"go.uber.org/atomic"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestAsyncQueue_Run(t *testing.T) {
+	q := newAsyncQueue(10, 10)
+	defer q.stop()
+
+	var i atomic.Int32
+	_ = q.run(func() { i.Add(1) })
+	_ = q.run(func() { i.Add(-1) })
+	_ = q.run(func() { i.Add(1) })
+	_ = q.run(func() { i.Add(-1) })
+	_ = q.run(func() { i.Add(1) })
+
+	// Wait for all operations to finish.
+	time.Sleep(100 * time.Millisecond)
+
+	require.Equal(t, int32(1), i.Load())
+}
+
+func TestAsyncQueue_QueueFullError(t *testing.T) {
+	const queueLength = 10
+
+	q := newAsyncQueue(queueLength, 1)
+	defer q.stop()
+
+	doneCh := make(chan struct{})
+	defer func() {
+		close(doneCh)
+	}()
+
+	// Keep worker busy.
+	_ = q.run(func() {
+		<-doneCh
+	})
+	time.Sleep(100 * time.Millisecond)
```

Comment:
Why is this required? Shouldn't be.

Reply:
Here we try to simulate a long running task that prevents our only worker from pulling any enqueued operation. This way we can fill the queue until getting the expected error. 

The sleep before the loop is done in order to ensure that the worker is already running before enqueuing any extra ops. Otherwise we may end up getting a _queue full error_ in the case it haven't triggered yet.